### PR TITLE
feat: improve notifications, interface scaling, and transcript UX

### PR DESCRIPTION
**Pull request by Yoonwoo-Ha**
## 배경

모바일·PWA에서 ChatMux를 사용할 때 확인된 다음 문제를 함께 개선합니다.

1. 완료 알림을 눌러도 해당 대화가 아니라 홈으로 이동하는 경우가 있었습니다.
2. 코드 에디터 글꼴 설정이 실제 편집기 전체에 반영되지 않았고, 채팅·사이드바·설정 화면의 전체 글꼴 크기는 조절할 수 없었습니다.
3. ChatMux에서 Codex 응답을 중단하면 turn만 멈추는 것이 아니라 Codex CLI 프로세스까지 종료될 수 있었습니다.
4. 이미지가 포함된 긴 transcript는 미리보기를 사용하지 않아도 이미지 정보를 전송했고, 대용량 Codex JSONL은 페이지를 열 때마다 처음부터 다시 읽었습니다.
5. 대용량 도구 출력이 standalone `tool_result`와 tool card 안에 중복으로 포함되어 응답 크기와 렌더링 비용이 커졌습니다.

### 알림 딥링크

- 외부 CLI 완료 알림 URL에 내부 alias가 아닌 실제 ChatMux `appSessionId`를 사용합니다.
- 아직 앱 세션으로 매핑되지 않은 generation은 존재하지 않는 세션 URL 대신 `/`로 이동합니다.
- service worker가 `/session/:id`의 세션 ID를 보존합니다.
- 알림 클릭 시 열린 클라이언트에 `notification:navigate`를 먼저 보내 SPA 라우팅을 수행합니다.
- iOS에서 `WindowClient.navigate()`가 거부되어도 클릭 처리가 실패하지 않도록 보조 경로를 유지합니다.

### 인터페이스 및 에디터 글꼴

- CodeMirror 본문과 gutter에 에디터 글꼴 크기를 명시적으로 적용합니다.
- 기본 에디터 글꼴 크기를 외관 설정과 같은 14px로 맞춥니다.
- 외관 설정에 전체 인터페이스 글꼴 크기를 추가합니다.
  - 작게: 14px / 87.5%
  - 중간: 16px / 100%
  - 크게: 18px / 112.5%
- 채팅, 세션 목록, 설정, 버튼과 상태 배지가 같은 비율로 조절됩니다.
- 설정값은 `localStorage`에 저장하고 React 렌더링 전에 적용해 초기 깜박임을 줄입니다.
- 10개 지원 언어에 설정 문구를 추가합니다.

### Codex 중단 동작

- Codex 외부 CLI의 정지 동작은 `Ctrl+C` 대신 `Escape`를 전달해 현재 turn만 중단합니다.
- 다른 provider는 기존 `Ctrl+C` 동작을 유지합니다.
- Codex `turn_aborted`를 사용자 입력 대기 상태로 정규화해 중단 후 잘못 RUN으로 남지 않게 합니다.

### 이미지 미리보기 설정

- 외관 설정에 채팅 이미지 미리보기 on/off를 추가합니다.
- OFF에서는 이미지 컴포넌트를 mount하지 않아 파일 및 blob 요청을 만들지 않습니다.
- history API에도 `includeImages=false`를 전달해 이미지 attachment와 base64 데이터를 응답에서 제외합니다.

### 대용량 transcript 성능

- Codex history cache를 raw JSONL 크기가 아닌 실제 normalized message 크기로 제한합니다.
- 96MB LRU 한도 안에서 큰 rollout도 유지해 같은 파일을 매 요청마다 다시 읽지 않습니다.
- 다음 요청에서는 기존 byte offset 이후의 append만 읽습니다.
- 파일 교체·truncate는 inode, 크기와 4KB boundary 비교로 감지해 캐시를 안전하게 재구축합니다.
- standalone `tool_result`는 제거하고 대응하는 tool-use card에만 결합합니다.
- 페이지 `total`, `limit`, `offset`이 실제 표시 메시지와 동일한 기준을 사용합니다.
- 64KB가 넘는 도구 출력은 앞·뒤 미리보기만 전송합니다.
- 사용자가 `전체 출력 불러오기`를 누르면 별도 endpoint에서 원문을 지연 로드합니다.

## 호환성과 안전성

- 기본 글꼴은 중간, 이미지 미리보기는 ON이므로 기존 기본 화면은 유지됩니다.
- 서버 DB 스키마와 migration 변경은 없습니다.
- 도구 출력 원문과 transcript 파일은 수정하거나 삭제하지 않습니다.
- 전체 출력은 명시적으로 요청할 때만 전송됩니다.
- Codex 캐시는 최대 4개 entry와 normalized 96MB LRU로 제한됩니다.
- 서버 재시작 뒤 첫 접근은 파일을 한 번 읽지만 이후 요청은 증분 처리됩니다.

## 검증

- `npm run typecheck` 통과
- `npm run lint` 통과
- `npm run build` 통과: client, server, Rust core
- 클라이언트 테스트 260개 통과
- Codex history 관련 테스트 13개 통과
- 실제 97MB Codex rollout
  - 기존: 요청마다 약 1.1~1.4초
  - 변경: 첫 요청 약 1.20초, 이후 약 0.002~0.003초
- 대용량 도구 출력 세션
  - 기존 history 응답 약 8.9MB
  - 변경 history 응답 약 220KB
  - 축약된 2MB 원문을 전체 출력 endpoint에서 정상 로드 확인
- 로컬 release 빌드를 3001 포트에 배포해 health, 페이지네이션, 이미지 제외, 전체 출력 로드를 확인했습니다.

### CODE REVIEW

File: server/modules/providers/list/codex/codex-sessions.provider.ts [hunk 1]
```diff
@@ -297,11 +301,43 @@ function parseCodexHistoryLine(line: string, accumulator: CodexHistoryAccumulato
 function touchCodexHistoryCache(sessionId: string, entry: CodexHistoryCacheEntry): void {
   codexHistoryCache.delete(sessionId);
   codexHistoryCache.set(sessionId, entry);
-  while (codexHistoryCache.size > CODEX_HISTORY_CACHE_MAX_ENTRIES) {
+  const cachedBytes = () => Array.from(codexHistoryCache.values())
+    .reduce((total, candidate) => total + candidate.normalizedBytes, 0);
+  while (
+    codexHistoryCache.size > CODEX_HISTORY_CACHE_MAX_ENTRIES
+    || (codexHistoryCache.size > 1 && cachedBytes() > CODEX_HISTORY_CACHE_MAX_NORMALIZED_BYTES)
```
Comment (devswha) [inline server/modules/providers/list/codex/codex-sessions.provider.ts]: **P1 — cache budget is not a bound for one oversized entry.** The byte eviction condition stops once `size === 1`, so a transcript whose normalized graph exceeds 96 MiB remains cached indefinitely (and an oversized newest entry evicts every other entry first). Allow the current result to be returned uncached, or evict it too when it exceeds the budget.

File: server/modules/providers/list/codex/codex-sessions.provider.ts [hunk 2]
```diff
@@ -421,16 +470,11 @@ async function refreshCodexHistoryCache(
...
     entry.malformed = appended.malformed;
-    entry.modifiedAtMs = metadata.mtimeMs;
   }
 
-  if (entry.retainedBytes + Buffer.byteLength(entry.tail) <= CODEX_HISTORY_CACHE_MAX_RETAINED_BYTES) {
-    touchCodexHistoryCache(sessionId, entry);
-  } else {
-    codexHistoryCache.delete(sessionId);
-  }
+  touchCodexHistoryCache(sessionId, entry);
```
Comment (devswha) [inline server/modules/providers/list/codex/codex-sessions.provider.ts:509]: **P1 — the unterminated tail is unaccounted retained memory.** `entry.tail` can contain an arbitrarily large partial JSONL record while `normalizedBytes` remains zero; `touchCodexHistoryCache` then caches it. Include `Buffer.byteLength(entry.tail)` in retained-byte accounting and decline to cache entries over budget.

File: server/modules/providers/list/codex/codex-sessions.provider.ts [hunk 3]
```diff
@@ -369,10 +418,12 @@ async function refreshCodexHistoryCache(
     && entry.filePath === sessionFilePath
     && entry.device === metadata.dev
     && entry.inode === metadata.ino;
-  const appendOnly = entry != null
+  const boundaryMatches = entry != null
     && sameFile
     && metadata.size >= entry.offset
-    && !(metadata.size === entry.offset && metadata.mtimeMs !== entry.modifiedAtMs);
+    && (entry.offset === 0
+      || (await readCodexHistoryBoundary(sessionFilePath, entry.offset)).equals(entry.boundary));
```
Comment (devswha) [inline server/modules/providers/list/codex/codex-sessions.provider.ts:455]: **P2 — same-length in-place rewrites before this 4 KiB boundary stay stale.** Path/inode/size and the EOF boundary all still match if an earlier record is rewritten without changing length, so cached messages/token usage can be served indefinitely. Restore metadata invalidation (`ctimeMs` is safer than restorable `mtimeMs`) or fingerprint the cached prefix. The new test does not cover this: `first-` → `next-` shrinks the file by one byte and only exercises truncation.

File: server/modules/providers/services/sessions.service.ts [hunk 4]
```diff
@@ -18,6 +18,84 @@ type CreateAppSessionResult = {
...
+  const bytes = Buffer.byteLength(serialized);
+  if (bytes <= HISTORY_TOOL_OUTPUT_PREVIEW_BYTES) {
+    return { content, truncated: false, bytes };
+  }
+
+  const source = Buffer.from(serialized);
+  const headBytes = Math.floor(HISTORY_TOOL_OUTPUT_PREVIEW_BYTES * 0.75);
+  const tailBytes = HISTORY_TOOL_OUTPUT_PREVIEW_BYTES - headBytes;
+  const head = source.subarray(0, headBytes).toString('utf8');
+  const tail = source.subarray(source.length - tailBytes).toString('utf8');
```
Comment (devswha) [inline server/modules/providers/services/sessions.service.ts]: **P2 — fixed byte cuts can corrupt UTF-8 previews.** Either subarray boundary can split a multibyte Korean/CJK code point, and independent `toString("utf8")` calls emit U+FFFD despite valid persisted output. Move each cut to a valid UTF-8 boundary before decoding (and add a multibyte regression case).

File: public/sw.js [hunk 5]
```diff
@@ -179,8 +188,21 @@ self.addEventListener('notificationclick', event => {
 
   const navigation = navigationHref(event.notification.data?.navigation);
   if (navigation) {
+    const sessionId = event.notification.data?.navigation?.sessionId || null;
```
Comment (devswha) [inline public/sw.js]: **P2 — outstanding notifications lose the new iOS fallback after a worker upgrade.** Notifications created by the previous worker contain only `navigation.href`, so this resolves `sessionId` to null. If `WindowClient.navigate()` rejects (the iOS case this postMessage path addresses), AppContent receives null and routes to `/`. Derive the ID from the already validated `navigation` href when the stored field is absent.

File: src/components/chat/hooks/useChatSessionState.ts [hunk 6]
```diff
@@ -627,6 +631,7 @@ export function useChatSessionState({
     sessionStore.fetchFromServer(selectedSessionId, {
       limit: MESSAGES_PER_PAGE,
       offset: 0,
+      includeImages: showImagePreviews,
```
Comment (devswha) [inline src/components/chat/hooks/useChatSessionState.ts:684]: **P2 — re-enabling image previews does not restore cached attachments.** This dependency reruns the effect, but `shouldReplaceSessionMessageWindow(...)` returns false for the current cached session and exits at lines 585–592 before this fetch. A session loaded with previews off therefore keeps `images` absent until an unrelated refresh. Invalidate/refetch the current bounded window when the preference changes back to enabled.

**Review by devswha — CHANGES_REQUESTED** [review]
Requesting changes for six correctness/performance regressions. The focused Codex and service-worker tests pass (19/19), but they do not exercise these edge cases. No path traversal was found in the new full-tool-result route; default image behavior remains enabled.

File: server/modules/providers/list/codex/codex-sessions.provider.ts (hunk 1 above)
Comment (Yoonwoo-Ha) [reply server/modules/providers/list/codex/codex-sessions.provider.ts]: 수정했습니다. 단일 항목도 전체 캐시 예산을 넘으면 현재 요청에는 반환하되 캐시에는 넣지 않도록 변경했고, size === 1 예외도 제거했습니다. 작은 예산으로 단일 oversized 항목을 검증하는 회귀 테스트를 추가했습니다.

File: server/modules/providers/list/codex/codex-sessions.provider.ts (hunk 2 above)
Comment (Yoonwoo-Ha) [reply server/modules/providers/list/codex/codex-sessions.provider.ts:509]: 수정했습니다. normalizedBytes뿐 아니라 Buffer.byteLength(entry.tail)과 boundary buffer까지 실제 보유량에 포함합니다. tail만으로 예산을 넘는 경우도 캐시되지 않는 회귀 케이스를 추가했습니다.

File: server/modules/providers/list/codex/codex-sessions.provider.ts (hunk 3 above)
Comment (Yoonwoo-Ha) [reply server/modules/providers/list/codex/codex-sessions.provider.ts:455]: 수정했습니다. 같은 inode/size일 때 ctimeMs가 달라지면 append-only 캐시를 무효화하고 처음부터 다시 파싱합니다. 기존 테스트도 first- → later-로 실제 동일 길이가 되게 고쳤고, 교체 전후 파일 바이트 크기가 같음을 명시적으로 검증하면서 mtime은 복원합니다.

File: server/modules/providers/services/sessions.service.ts (hunk 4 above)
Comment (Yoonwoo-Ha) [reply server/modules/providers/services/sessions.service.ts]: 수정했습니다. head/tail 절단 지점을 UTF-8 continuation byte 밖으로 이동한 뒤 디코딩하고, 실제 생략된 바이트 수도 조정된 경계로 계산합니다. 고정 48 KiB 경계가 한글 코드포인트 중간에 걸리는 회귀 테스트에서 U+FFFD가 없고 전체 결과 조회는 원문과 동일함을 확인합니다.

File: public/sw.js (hunk 5 above)
Comment (Yoonwoo-Ha) [reply public/sw.js]: 수정했습니다. 저장된 navigation.sessionId가 없으면 이미 same-origin 검증을 통과한 navigation href에서 세션 ID를 복원합니다. 이전 워커 형식의 payload와 WindowClient.navigate() 거절을 함께 재현하는 테스트를 추가했습니다.

File: src/components/chat/hooks/useChatSessionState.ts (hunk 6 above)
Comment (Yoonwoo-Ha) [reply src/components/chat/hooks/useChatSessionState.ts:684]: 수정했습니다. 현재 세션에서 이미지 미리보기가 false → true로 바뀌면 기존 메시지/스크롤 창을 초기화하지 않고 refreshFromServer(includeImages: true)로 현재 로드된 bounded window만 백그라운드 재조회합니다. 동일 세션·캐시 존재·재활성화 조건을 고정하는 회귀 테스트도 추가했습니다.

**Comment by Yoonwoo-Ha** [comment]
리뷰 6건을 `c5383a4`에 모두 반영했습니다.

- Codex 캐시: 단일 oversized 항목 비캐시, tail/boundary 용량 포함, 동일 길이 덮어쓰기 ctime 무효화
- 전송 미리보기: UTF-8 안전 경계 절단 및 한글 회귀 테스트
- 알림: 구형 서비스 워커 payload의 세션 ID fallback
- 이미지: 미리보기 재활성화 시 현재 bounded window만 백그라운드 재조회

검증 결과:
- 관련 서버/클라이언트 집중 회귀: 35/35 통과
- typecheck, lint, build, check:core(19/19), check:identity 통과
- GitHub CI Verify Node 22 / Node 24 모두 통과

각 리뷰 스레드에도 변경 내용과 회귀 케이스를 답변하고 resolved 처리했습니다. 재검토 부탁드립니다.

**Review by devswha — APPROVED** [review]
Re-reviewed head 17ad2b0. The cached image refresh now fences stale session results, all ten locales contain the full-tool-output strings, the new regression test passes, and the full local npm run verify gate completed successfully.
